scipy/optimize/_gensa.py

```diff
@@ -0,0 +1,601 @@
+            if task_str.startswith(b'FG'):
+                self._xbuffer = np.array(x)
+                f = self._fobjective(self._xbuffer)
+                if self._knowreal:
+                    if f <= self._realthreshold:
+                        self._xbuffer = np.array(x)
+                        self._fvalue = f
+                        return
+                g = np.array(self._compute_gradient(x), np.float64)
+            elif task_str.startswith(b'NEW_X'):
+                pass
+            else:
+                self._fvalue = f
+                self._xbuffer = np.array(x)
+                return
+
+    def _compute_gradient(self, x):
+        """Computation of numerical derivatives for local search"""
+        # Can be done later here:
+        # Check here if the expression of the derivatives is provided
+        # For that, add an argument for jacobian in the main function
+        # and use MemoizeJac from optimize.py
+        g = np.zeros(self._dim, np.float64)
+        for i in range(self._dim):
+            x1 = np.array(x)
+            x2 = np.array(x)
+            respl = self._reps
+            respr = self._reps
+            x1[i] = x[i] + respr
+            if x1[i] > self._upper[i]:
+                x1[i] = self._upper[i]
+                respr = x1[i] - x[i]
+            x2[i] = x[i] - respl
+            if x2[i] < self._lower[i]:
+                x2[i] = self._lower[i]
+                respl = x[i] - x2[i]
+            f1 = self._fobjective(x1)
+            f2 = self._fobjective(x2)
+            g[i] = ((f1 - f2)) / (respl + respr)
+        idx = np.logical_or(np.isnan(g), np.isinf(g))
+        g[idx] = 101.0
+        return g
+
+    # @profile
+    def _yygas(self):
+        if self._usey == 1:
+            enter = True
+            while(enter or (self._sgas <= 0 or self._sgas >= 1)):
+                enter = False
+                self._xgas = np.random.ranf() * 2.0 - 1.0
+                self._ygas = np.random.ranf() * 2.0 - 1.0
+                self._sgas = self._xgas * self._xgas + self._ygas * self._ygas
+            root = np.sqrt(-2.0 / self._sgas * np.log(self._sgas))
+            self._ranbyx = self._xgas * root
+            self._ranbyy = self._ygas * root
+            retval = self._ranbyy
+            self._usey = 0
+        else:
+            retval = self._ranbyx
+            self._usey = 1
+        return retval
+
+    @property
+    def result(self):
+        """ The OptimizeResult """
+        res = OptimizeResult()
+        res.x = self._xmini
+        res.fun = self._fvalue
+        res.message = self._message
+        res.nit = self._step_record
+        return res
+
+def gensa(func, x0, bounds, niter=500, T=5230., visitparam=2.62,
+        acceptparam=-5.0, maxtime=3600, maxcall=1e7, args=()):
+    """
+    Find the global minimum of a function using the Generalized Simulated
+    Annealing algorithm
+
+    Parameters
+    ----------
+    func : callable ``f(x, *args)``
```

Comment:
For the parameter list documentation I would just copy from another minimizer, for familiarity. The following is from `differential_evolution`.

    The objective function to be minimized.  Must be in the form
    ``f(x, *args)``, where ``x`` is the argument in the form of a
     1-D array and ``args`` is a  tuple of any additional fixed
     parameters needed to completely specify the function.